### ISSUE TITLE
Remote videos slow + ANR on Android 9 and 10

### ISSUE THREAD

**Issue by cristianoccazinsp** [TheWidlarzGroup/react-native-video#1853]
# Bug

On Android 9 and 10, remotely loaded videos seem to take a very long time to play. Seeking is also extremely slow (changing the seek position will result in the same initial load time).

What's worse, an ANR can happen while loading. Here's the ANR stack trace:

```
ANR ·MainActivity
Input dispatching timed out (Waiting to send non-key event because the touched window has not finished processing....

MediaPlayer.java:-2 · android.media.MediaPlayer.setPlaybackParams
--
ReactVideoView.java:470 · com.brentvatne.react.ReactVideoView.setRateModifier
ReactVideoView.java:521 · com.brentvatne.react.ReactVideoView.applyModifiers
ReactVideoView.java:561 · com.brentvatne.react.ReactVideoView.onPrepared
MediaPlayer.java:3346 · android.media.MediaPlayer$EventHandler.handleMessage


```

## Platform

Which player are you experiencing the problem on:
* Android MediaPlayer

## Environment info

React Native: 0.61.5
Library version: "react-native-video": "5.0.2"

## Steps To Reproduce

Load <Video> with the HTTP link provided below. Use on Android 9 or 10 (works fine on Android < 9).

...

## Video sample
If possible, include a link to the video that has the problem that can be streamed or downloaded from.
https://zinspectordev2.s3.amazonaws.com/usi/2/1576687667933fbdddc5f5458c8b013fef64127c99.mp4

**Comment by fabOnReact** [comment]
I have the same problem and I look forward in seeing it solved. I have optimized videos for the [width of the device](https://github.com/fabriziobertoglio1987/surf-react-native/blob/master/app/components/index/Cameras.js#L96).

they [are just few seconds long](https://res.cloudinary.com/dhhzzcjq0/video/upload/f_mp4,w_300/v1574808115/eqnqkux2nsjswfs9x6ey.mp4) and their size is something like 130kb 

but the [app (you can download it and test)](https://surfcheck.xyz) will take several seconds to load them

**Comment by cristianoccazinsp** [comment]
I solved the issue by manually linking the Android version and using the "exoplayer" version instead. That one works fine; I wonder why it is not the default implementation.

**Comment by fabOnReact** [comment]
I will try it too. Thanks

https://github.com/react-native-community/react-native-video#android-installation

**Comment by pyeonjaesik** [comment]
@cristianoccazinsp hi!
how did you link exoplayer???

**Comment by cristianoccazinsp** [comment]
@pyeonjaesik  I just added the required gradle changes (pretty much manual linking). Here's what I did with RN 0.61 (taken from https://github.com/react-native-community/react-native-video#android-installation)

First add/update `react-native.config.js` to tell RN to not auto link the module:
```
module.exports = {
  dependencies: {
    // video we manually link on Android to use exoplayer
    'react-native-video': {
      platforms: {
        android: null
      },
    }
  },
};
```

Then update `settings.gradle` to add the dependency:
```
include ':react-native-video'
project(':react-native-video').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-video/android-exoplayer')
```

Also update `build.gradle`:

```
dependencies {
    implementation project(':react-native-video')
    ... others ...
}
```

Add package to `MainApplication.java`:
```
import com.brentvatne.react.ReactVideoPackage;

... other code...

@Override
    protected List<ReactPackage> getPackages() {
      @SuppressWarnings("UnnecessaryLocalVariable")
      List<ReactPackage> packages = new PackageList(this).getPackages();
     
      packages.add(new ReactVideoPackage());

      return packages;
    }

```

I probably missed something, but the instructions are clear on the link above.

**Comment by pyeonjaesik** [comment]
@cristianoccazinsp  Thank you! I solved this problem. Thanks to your help, remote loading is faster than before. But I want  far more faster than this . Can you give me a tip?

**Comment by cristianoccazinsp** [comment]
I don't think you can make it even faster.  You may try fine tuning the library settings, or looking at your backend to see if you're providing a good/fast stream. Other than that, I don't know :(

As a last resource, try a webview with a youtube player :)

**Comment by pyeonjaesik** [comment]
what do you mean "You may try fine tuning the library settings" ?

**Comment by cristianoccazinsp** [comment]
I think there are various buffering and stream setting you can test, but I
haven't tried that at all.

**Comment by swizes** [comment]
@cristianoccazinsp did you use samsung phone while uploading the video? Based on my experience, that slowness happens when a samsung device uploads a video, remote mp4 urls etc. are working fine

**Comment by cristianoccazinsp** [comment]
Nope, Pixel 2.

**Comment by swizes** [comment]
I would like to confirm, how long does it take on your app to load the video? The problem is, android loads videos fast, if the video is uploaded from iphone... I dont understand what causes the issue

**Comment by cristianocca** [comment]
Using Exoplayer, videos load acceptably fast; about 2 or 3 seconds of buffering. Note that our videos come from AWS S3, which by itself has a some latency. Rather than that, I don't see difference between devices. Your Samsung is probably uploading much bigger videos than your iPhone.

**Comment by maxime-memtell** [comment]
TLDR; Oh my... Without being too certain, I think that fixed my issue of multiple videos on one page (via snap-carousel) not starting/being stuck on black screen.

Additional infos for anyone finding this later:
* I say I'm not too sure because of the fact that the issue was not reliably happening, sometimes yes sometimes no. Definitely never on iOS
* I'm using https://github.com/meliorence/react-native-snap-carousel with a video on each slide. The problem was either none or all videos at the same time would take forever to load
* "stuck on black screen" means it would go into buffer mode ({"isBuffering": true}) and stay there, while displaying a black screen. When waiting a looooong time (several minutes), the videos would eventually start. My videos are 10s long, ~3M each, 4 at a time and hosted on S3. It's definitely fast when this bug doesn't happen

Update:
Nope, didn't fix it... If I navigate through another screen before (@react-navigation), then it can still happened. I added another fix on top where I rendered the video on the previous screen only if it was in focus.
I think I'll do the same everywhere I'm using the <Video/> component from there on. Cannot be too bad anyway since I know react-navigation doesn't unmount screens in the stack and I'm using a lot of videos throughout the app!

**Comment by alexghi** [comment]
@maxime-memtell could you please tell me how you managed to get it done? I'm facing very long load times for my audio streams on Android (never had any issue on ios) but on Android I can't seem to make it load fast enought to make it production-ready
